### ISSUE TITLE
Context aware templates

### ISSUE THREAD

**Issue by agentgt** [jstachio/jstachio#168]
As @dsyer found out through their experiments with Spring integration it is still a fairly big pain when dealing with cross cutting attributes.

Cross cutting attributes would be things like CSRF token.

The documentation (overview.html) discusses some remedies on this and the Spring examples show some solutions.

The two solutions quickly are:

 1. Threadlocals or similar
 1. The model has a mutable attribute that gets set right before rendering

The threadlocal solution is a nonstarter for any reactive based web application (google if you don't know why that is a problem). 

Furthermore even for traditional blocking applications it is unclear if it is even a long term solution with the advent of project loom which is now basically in the JDK.

The second solution is undesirable in that the model now has to know about cross cutting crap and thus you get a pattern where every model class needs a mutable global context as some collaborator (or the model is mutable).

This is rather painful if one wants to use immutable `record`s for models.

```java
public record Blog(Context ctx, String title, ...) implements ContextHolder {}
public record Home(Context ctx, String title, ...) implements ContextHolder {}

public Home someController(...) {
    Context ctx = Context.notReady(); // will be filled at end of request
    return new Home(ctx, ....);
}
```

For an advance DIY Java developer the above is probably not that painful and can be mitigated using all sorts of Java OOP / functional techniques. But for an entry level Spring Boot developer probably not and there is no real best practice other than what is provided in the Spring examples.

**Perhaps we can do a JStachio solution that can be re-used across web applications.**

Perhaps we can generate templates code that will take a special interface called `Context` (the name is a placeholder and we bikeshed that later).

So instead of:

```
    public <A extends Output<E>, E extends Exception> A execute(T model, A output) throws E;
```

We have an additional Template interface that has an additional method:

```
    public <A extends Output<E>, E extends Exception> A execute(T model, A output, Context ctx) throws E;
```

`Context` is a functional interface that is basically `Function<String, @Nullable String>`.

An empty `Context` for when it is not passed would simply be `(String s) -> null`.

Then templates will have a virtual binding that we will call for now `@context` (perhaps this could be configurable... again bikeshed) that essentially appears to be like a `Map<String,String>`.

```java
@JStache(template = """
{{title}} {{! this is on the model }} 
<input type="hidden" name="_csrf" value="{{@context.csrf}}" /> {{! this is not }}
""")
public record Blog(String title){}
```

Then web frameworks call the `execute` method with their own custom `Context` (for servlet based environments this could be as simple as `request::getAttribute`).

**Comment by dsyer** [comment]
I like the idea of a `Context` that is `Function<String, String>`. I think that probably avoids some of the ugliness of (for example) Rocker. It still opens JStachio up to runtime errors when a context entry can't be found? I suppose we are trading convenience for that risk.

If I'm reading the sample right above you are proposing a naming convention with `@context` as a prefix for the renderer? I can see possibilities there to extend the convention to `@other_things` via an extension point. Or is that over thinking it, and we don't need the "context" part of the prefix (i.e. `{{@csrf}}` would be sufficient)?

In any case the proof of the pudding will be in the eating and I'm happy to try it out as soon as you have a proposal, or clarifications on the above.

**Comment by SerayaEryn** [comment]
This would be a very useful feature. handlebars uses `@root` access the root/global context. 

Another feature that is maybe related to this is accessing the parent context (`../example` in handlebars).

**Comment by agentgt** [comment]
As I'm doing more integrations I'm seeing how supporting this will be challenging given how more and more frameworks don't seem to expose the request or the context in their "view" or "template engine" SPIs.

For example Dropwizard and Micronaut you cannot access the request easily when at the rendering state.

https://micronaut-projects.github.io/micronaut-views/latest/api/io/micronaut/views/ViewsRenderer.html

**Comment by agentgt** [comment]
This change will seal off `ContextNode`. Before it was open and anyone could implement the interface. It is too dangerous to leave it open so it is now `sealed`.

The API documentation did warn about the dangers of `ContextNode` and I seriously doubt anyone implemented it or is using it so I am semi violating semver policy.

**Comment by agentgt** [comment]
@dsyer @SerayaEryn context support is now in main and will be released in 1.3.0.

@dsyer I hid the ContextAwareOutput and instead you can use `ContextJStachio.of(jstachio)`.

Most JStachio will be `ContextJStachio` so the above is very fast and no allocation.

In 2.0 I will consider moving it into the regular JStachio interface.

**Comment by agentgt** [comment]
Another breaking change is JStachioFilter but to be honest it will probably get removed in 2.0 and the doc warned how it was not really public.
